Add `const` support for float rounding methods

ruancomelli opened the pull request:
# Add `const` support for float rounding methods

This PR makes the following float rounding methods `const`:

- `f64::{floor, ceil, trunc, round, round_ties_even}`
- and the corresponding methods for `f16`, `f32` and `f128`

Tracking issue: https://github.com/rust-lang/rust/issues/141555

## Procedure

I followed https://github.com/rust-lang/rust/commit/c09ed3e767a73d83673790f74c357432fa44d320 as closely as I could in making float methods `const`, and also received great guidance from https://internals.rust-lang.org/t/const-rounding-methods-in-float-types/22957/3?u=ruancomelli.

## Note

This is my first code contribution to the Rust project, so please let me know if I missed anything - I'd be more than happy to revise and learn more. Thank you for taking the time to review it!

ruancomelli committed:
Add const support for float rounding methods

Add const support for the float rounding methods floor, ceil, trunc,
fract, round and round_ties_even.
This works by moving the calculation logic from

     src/tools/miri/src/intrinsics/mod.rs

into

     compiler/rustc_const_eval/src/interpret/intrinsics.rs.

All relevant method definitions were adjusted to include the `const`
keyword for all supported float types: f16, f32, f64 and f128.

The constness is hidden behind the feature gate

     feature(const_float_round_methods)

which is tracked in

     https://github.com/rust-lang/rust/issues/141555

This commit is a squash of the following commits:
- test: add tests that we expect to pass when float rounding becomes const
- feat: make float rounding methods `const`
- fix: replace `rustc_allow_const_fn_unstable(core_intrinsics)` attribute with `#[rustc_const_unstable(feature = "f128", issue = "116909")]` in `library/core/src/num/f128.rs`
- revert: undo update to `library/stdarch`
- refactor: replace multiple `float_<mode>_intrinsic` rounding methods with a single, parametrized one
- fix: add `#[cfg(not(bootstrap))]` to new const method tests
- test: add extra sign tests to check `+0.0` and `-0.0`
- revert: undo accidental changes to `round` docs
- fix: gate `const` float round method behind `const_float_round_methods`
- fix: remove unnecessary `#![feature(const_float_methods)]`
- fix: remove unnecessary `#![feature(const_float_methods)]` [2]
- revert: undo changes to `tests/ui/consts/const-eval/float_methods.rs`
- fix: adjust after rebase
- test: fix float tests
- test: add tests for `fract`
- chore: add commented-out `const_float_round_methods` feature gates to `f16` and `f128`
- fix: adjust NaN when rounding floats
- chore: add FIXME comment for de-duplicating float tests
- test: remove unnecessary test file `tests/ui/consts/const-eval/float_methods.rs`
- test: fix tests after upstream simplification of how float tests are run

RalfJung commented:
> However, I could have implemented a single one taking a rustc_apfloat::Round parameter:

Yes, that's a lot better I think. :)

RalfJung commented:
Note that your branch has conflicts with the master branch. Not sure how that happened given that the PR is fresh, but it means you'll have to [rebase and resolve those conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

ruancomelli commented:
Strange, CI is failing with [intrinsic <X> is not supported at compile-time](https://github.com/rust-lang/rust/actions/runs/15231349857/job/42839564438?pr=141521) for each one of the new methods.
I'm looking into that.

ruancomelli commented:
As far as I can see, the main difference between the newly `const` methods (e.g. `floor`) and the ones that already work (e.g. `abs`) is the attributes. So my current hypotheses is that I got those wrong.

CAD97 commented on library/core/src/intrinsics/mod.rs:
```diff
@@ -2294,10 +2294,11 @@ pub fn round_ties_even_f16(x: f16) -> f16;
 /// Provided for compatibility with stdarch. DO NOT USE.
 #[inline(always)]
+#[rustc_allow_const_fn_unstable(core_intrinsics)]
 pub unsafe fn rintf32(x: f32) -> f32 {
```
I think you meant to mark this as `const` as well?

RalfJung replied:
I don't think the `rintf` functions need to be constified at all.

CAD97 commented on library/core/src/num/f32.rs:
```diff
@@ -1591,7 +1591,8 @@ pub mod math {
     #[unstable(feature = "core_float_math", issue = "137578")]
     #[must_use = "method returns a new number and does not mutate the original value"]
-    pub fn floor(x: f32) -> f32 {
+    #[rustc_allow_const_fn_unstable(core_intrinsics)]
```
The reason you need this is that you're "stably" using functionality which is marked unstably const, IIUC. You probably want to use

```rust
    #[rustc_const_unstable(feature = "const_float_methods", issue = "130843")]
```

here just like you did in std.

RalfJung replied:
Indeed, definitely do *not* use `rustc_allow_const_fn_unstable`.

ruancomelli replied:
@CAD97 it looks like `const_float_methods` was stabilize; in this file we have a couple of occurrences of `#[rustc_const_stable(feature = "const_float_methods", since = "1.85.0")]`. In fact, if I try `#[rustc_const_unstable(feature = "const_float_methods", issue = "130843")]`, I get the following error:

```
error[E0711]: feature `const_float_methods` is declared unstable, but was previously declared stable
    --> library/core/src/num/f32.rs:1594:5
     |
1594 |     #[rustc_const_unstable(feature = "const_float_methods", issue = "130843")]
     |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

For more information about this error, try `rustc --explain E0711`.
```

@RalfJung I picked this decorator because of the following compile error that happened without adding any attributes:

```
error: const function that might be (indirectly) exposed to stable cannot use `#[feature(core_intrinsics)]`
    --> library/core/src/num/f32.rs:1597:18
     |
1597 |         unsafe { intrinsics::floorf32(x) }
     |                  ^^^^^^^^^^^^^^^^^^^^^^^
     |
help: if the function is not (yet) meant to be exposed to stable, add `#[rustc_const_unstable]` (this is what you probably want to do)
     |
1595 +     #[rustc_const_unstable(feature = "...", issue = "...")]
1596 |     pub const fn floor(x: f32) -> f32 {
     |
help: otherwise, as a last resort `#[rustc_allow_const_fn_unstable]` can be used to bypass stability checks (this requires team approval)
     |
1595 +     #[rustc_allow_const_fn_unstable(core_intrinsics)]
1596 |     pub const fn floor(x: f32) -> f32 {
     |

error: could not compile `core` (lib) due to 1 previous error
```

The attribute `#[rustc_const_unstable(feature = "core_float_math", issue = "137578")]` works fine. I didn't find any other relevant tracking issues for the `const`-ness of float methods.
I don't fully understand the process regarding stabilization of features in the compiler; would `#[rustc_const_stable(feature = "const_float_methods", since = "1.85.0")]` be a good choice here? Note that these rounding functions are all decorated with `#[unstable(feature = "core_float_math", issue = "137578")]`.

RalfJung replied:
Since this is a new unstable feature, you should come up with a new name for it -- such as `const_float_round`. Then you use that name for each of the functions here. You should also create a tracking issue (use the "library tracking issue" template), and then put the issue number into the attribute as well.

This is probably explained somewhere in the dev guide but I don't know where.^^ Let me know if you have further questions. :)

CAD97 commented on library/std/src/f64.rs:
```diff
@@ -66,17 +67,13 @@ impl f64 {
+    pub const fn ceil(self) -> f64 {
         core::f64::math::ceil(self)
     }
-
-    /// Returns the nearest integer to `self`. If a value is half-way between two
-    /// integers, round away from `0.0`.
-    ///
```
This deletion looks unintended.

CAD97 reviewed (commented):
AIUI, the reason you're hitting "intrinsic not supported at compile time" is that std currently gets built with the bootstrap compiler, which is then used when building the new compiler. So you need to have a `#[cfg(bootstrap)]` version of the function/intrinsic which isn't `const` and a `#[cfg(not(bootstrap))]` which is. 

IIUC, we're in the process of flipping this so that the compiler gets built with the bootstrap std, and std always gets built with the new compiler. Once that switch happens, this won't be necessary anymore.

RalfJung commented:
> AIUI, the reason you're hitting "intrinsic not supported at compile time" is that std currently gets built with the bootstrap compiler, which is then used when building the new compiler. So you need to have a #[cfg(bootstrap)] version of the function/intrinsic which isn't const and a #[cfg(not(bootstrap))] which is.

The problem occurs when building the tests. So you don't need to do any `cfg` in libcore itself, but the new tests in `library/coretests` need a `#[cfg(not(bootstrap))]`.

RalfJung commented on library/coretests/tests/num/mod.rs:
```diff
@@ -924,6 +924,77 @@ macro_rules! test_float {
+            #[test]
+            fn floor() {
+                $fassert!((0.0 as $fty).floor(), 0.0);
+                $fassert!((-0.0 as $fty).floor(), -0.0);
```
Note that `0.0 == -0.0`, so you are not actually testing that the zero has the right sign here.

Maybe the `float_assert` and `float_const_assert` macros should compare `$left.to_bits()` and `$right.to_bits()` instead?

ruancomelli replied:
That's a good point. I tried doing that, but it caused many compilation errors because e.g. `0.0` is of unknown float type. A solution for that would be to write `0.0 as $fty` in this entire file (including the other existing tests), which, in turn, caused some tests to fail (I think because of under- or overflows).

To avoid having this grow too much in scope, I followed the other tests' strategy by adding `.is_sign_positive()` and `.is_sign_negative()` tests. They allow us to check that the zero has the correct sign.

Please let me know if this strategy is OK, or if you'd still prefer the `.to_bits()` solution.

RalfJung commented on tests/ui/consts/const-eval/float_methods.rs:
No need to add two complete sets of tests.

RalfJung replied:
Yeah -- you're good on test coverage with the other tests. I should look into removing this file entirely and having library tests for it instead.

RalfJung commented:
Your functions are not meant to be stable, right? So do the part under "help: if the function is not (yet) meant to be exposed to stable". The other part explicitly says "as a last resort" and "requires team approval"... I  don't know how much more scary we have to make that message to stop people from using the attribute. :(

Maybe we should just stop mentioning it entirely.

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-19`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/15239269461/job/42857289667) [(plain)](https://github.com/rust-lang/rust/commit/f84737e23345c5b7905069660f5fc0ce67b11c9e/checks/42857289667/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
#18 exporting to docker image format
#18 sending tarball 21.0s done
#18 DONE 27.4s
##[endgroup]
Setting extra environment values for docker:  --env ENABLE_GCC_CODEGEN=1 --env GCC_EXEC_PREFIX=/usr/lib/gcc/
[CI_JOB_NAME=x86_64-gnu-llvm-19]
[CI_JOB_NAME=x86_64-gnu-llvm-19]
debug: `DISABLE_CI_RUSTC_IF_INCOMPATIBLE` configured.
---
sccache: Listening on address 127.0.0.1:4226
##[group]Configure the build
configure: processing command line
configure: 
configure: build.configure-args := ['--build=x86_64-unknown-linux-gnu', '--llvm-root=/usr/lib/llvm-19', '--enable-llvm-link-shared', '--set', 'rust.randomize-layout=true', '--set', 'rust.thin-lto-import-instr-limit=10', '--set', 'build.print-step-timings', '--enable-verbose-tests', '--set', 'build.metrics', '--enable-verbose-configure', '--enable-sccache', '--disable-manage-submodules', '--enable-locked-deps', '--enable-cargo-native-static', '--set', 'rust.codegen-units-std=1', '--set', 'dist.compression-profile=balanced', '--dist-compression-formats=xz', '--set', 'rust.lld=false', '--disable-dist-src', '--release-channel=nightly', '--enable-debug-assertions', '--enable-overflow-checks', '--enable-llvm-assertions', '--set', 'rust.verify-llvm-ir', '--set', 'rust.codegen-backends=llvm,cranelift,gcc', '--set', 'llvm.static-libstdcpp', '--set', 'gcc.download-ci-gcc=true', '--enable-new-symbol-mangling']
configure: build.build          := x86_64-unknown-linux-gnu
configure: target.x86_64-unknown-linux-gnu.llvm-config := /usr/lib/llvm-19/bin/llvm-config
configure: llvm.link-shared     := True
configure: rust.randomize-layout := True
configure: rust.thin-lto-import-instr-limit := 10
---
---- num::f32_const::floor stdout ----
---- num::f32_const::floor stderr ----

thread 'main' panicked at coretests/tests/num/mod.rs:1053:1:
assertion failed: const { (0.0 as f32).floor().is_sign_negative() }
stack backtrace:
   0: __rustc::rust_begin_unwind
   1: core::panicking::panic_fmt
   2: core::panicking::panic
   3: coretests::num::f32_const::floor
---
---- num::f32::floor stdout ----
---- num::f32::floor stderr ----

thread 'main' panicked at coretests/tests/num/mod.rs:1041:1:
assertion failed: (0.0 as f32).floor().is_sign_negative()
stack backtrace:
   0: __rustc::rust_begin_unwind
   1: core::panicking::panic_fmt
   2: core::panicking::panic
   3: coretests::num::f32::floor
---
---- num::f64::floor stdout ----
---- num::f64::floor stderr ----

thread 'main' panicked at coretests/tests/num/mod.rs:1065:1:
assertion failed: (0.0 as f64).floor().is_sign_negative()
stack backtrace:
   0: __rustc::rust_begin_unwind
   1: core::panicking::panic_fmt
   2: core::panicking::panic
   3: coretests::num::f64::floor
---
---- num::f64_const::floor stdout ----
---- num::f64_const::floor stderr ----

thread 'main' panicked at coretests/tests/num/mod.rs:1077:1:
assertion failed: const { (0.0 as f64).floor().is_sign_negative() }
stack backtrace:
   0: __rustc::rust_begin_unwind
   1: core::panicking::panic_fmt
   2: core::panicking::panic
   3: coretests::num::f64_const::floor
---
    num::f64_const::floor

test result: FAILED. 1902 passed; 4 failed; 4 ignored; 0 measured; 0 filtered out; finished in 1.29s

error: test failed, to rerun pass `-p coretests --test coretests`
env -u RUSTC_WRAPPER CARGO_ENCODED_RUSTDOCFLAGS="-Csymbol-mangling-version=v0\u{1f}-Zrandomize-layout\u{1f}-Zunstable-options\u{1f}--check-cfg=cfg(bootstrap)\u{1f}--check-cfg=cfg(llvm_enzyme)\u{1f}-Dwarnings\u{1f}-Wrustdoc::invalid_codeblock_attributes\u{1f}--crate-version\u{1f}1.89.0-nightly\t(95aa983b6\t2025-05-25)" CARGO_ENCODED_RUSTFLAGS="-Csymbol-mangling-version=v0\u{1f}-Zrandomize-layout\u{1f}-Zunstable-options\u{1f}--check-cfg=cfg(bootstrap)\u{1f}--check-cfg=cfg(llvm_enzyme)\u{1f}-Zmacro-backtrace\u{1f}-Csplit-debuginfo=off\u{1f}-Clink-arg=-L/usr/lib/llvm-19/lib\u{1f}-Cllvm-args=-import-instr-limit=10\u{1f}-Clink-args=-Wl,-z,origin\u{1f}-Clink-args=-Wl,-rpath,$ORIGIN/../lib\u{1f}-Alinker-messages\u{1f}--cap-lints=allow\u{1f}--cfg\u{1f}randomized_layouts" RUSTC="/checkout/obj/build/x86_64-unknown-linux-gnu/stage1-tools/cg_clif/dist/rustc-clif" RUSTDOC="/checkout/obj/build/x86_64-unknown-linux-gnu/stage1-tools/cg_clif/dist/rustdoc-clif" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/cargo" "test" "--manifest-path" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage1-tools/cg_clif/build/sysroot_tests/Cargo.toml" "--target-dir" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage1-tools/cg_clif/build/sysroot_tests_target" "--locked" "--target" "x86_64-unknown-linux-gnu" "-p" "coretests" "-p" "alloctests" "--tests" "--" "-q" exited with status ExitStatus(unix_wait_status(25856))
Command has failed. Rerun with -v to see more details.
Build completed unsuccessfully in 0:24:33
  local time: Sun May 25 15:48:48 UTC 2025
  network time: Sun, 25 May 2025 15:48:49 GMT
##[error]Process completed with exit code 1.
Post job cleanup.
```

</details>

ruancomelli commented:
Hey @RalfJung and @CAD97, I've addressed or replied to all of your review comments.

I've also created a tracking issue for this feature here: #141555.

I have two main open questions now.

First one is: should the rounding methods on `f16` and `f128` be gated behind `#[rustc_const_unstable(feature = {"f16", "f128"}, issue = "116909")]` (from #116909) or `#[rustc_const_unstable(feature = "const_float_round_methods", issue = "141555")]`? I think it should be the former one, but I'm not sure.

The second question: I added `#![feature(const_float_round_methods)]` to [`library/std/src/lib.rs`](https://github.com/rust-lang/rust/pull/141521/commits/805077f140be85968f33050bc1038eb620666e80#diff-317b30ba653f20600042781487d517ae7d07547466b074d09d521f0b9b6ccc52), which was suggested by the compiler itself. Without it, `./x check` results in the following errors:
```
error: `core::f32::math::floor` is not yet stable as a const fn
  --> library/std/src/f32.rs:50:9
   |
50 |         core::f32::math::floor(self)
   |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
help: add `#![feature(const_float_round_methods)]` to the crate attributes to enable
  --> library/std/src/lib.rs:433:1
   |
433+ #![feature(const_float_round_methods)]
   |

... and similar for the other rounding methods

error: could not compile `std` (lib) due to 12 previous errors
```
This is similar to [what was done in the reference PR](https://github.com/rust-lang/rust/pull/130568/files#diff-ea30f57e78f3b861b784ae315b3ac31358b7cea2aa14a668084ad412de12e586), even though they changed [library/core/src/lib.rs](https://github.com/rust-lang/rust/pull/130568/files#diff-ea30f57e78f3b861b784ae315b3ac31358b7cea2aa14a668084ad412de12e586) (notice `core` and not `std`).
However, I'm not sure if this is the correct way to proceed. I read the [dev guide book](https://std-dev-guide.rust-lang.org), but that question remained unanswered.

rust-log-analyzer commented:
The job **`mingw-check`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/15240157729/job/42859255019) [(plain)](https://github.com/rust-lang/rust/commit/805077f140be85968f33050bc1038eb620666e80/checks/42859255019/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain

```

</details>

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-19`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/15241188080/job/42861469230) [(plain)](https://github.com/rust-lang/rust/commit/4512002093572d2c218ae4d3fbf110bea7d9009f/checks/42861469230/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
#18 exporting to docker image format
#18 sending tarball 20.1s done
#18 DONE 27.2s
##[endgroup]
Setting extra environment values for docker:  --env ENABLE_GCC_CODEGEN=1 --env GCC_EXEC_PREFIX=/usr/lib/gcc/
[CI_JOB_NAME=x86_64-gnu-llvm-19]
[CI_JOB_NAME=x86_64-gnu-llvm-19]
debug: `DISABLE_CI_RUSTC_IF_INCOMPATIBLE` configured.
---
sccache: Listening on address 127.0.0.1:4226
##[group]Configure the build
configure: processing command line
configure: 
configure: build.configure-args := ['--build=x86_64-unknown-linux-gnu', '--llvm-root=/usr/lib/llvm-19', '--enable-llvm-link-shared', '--set', 'rust.randomize-layout=true', '--set', 'rust.thin-lto-import-instr-limit=10', '--set', 'build.print-step-timings', '--enable-verbose-tests', '--set', 'build.metrics', '--enable-verbose-configure', '--enable-sccache', '--disable-manage-submodules', '--enable-locked-deps', '--enable-cargo-native-static', '--set', 'rust.codegen-units-std=1', '--set', 'dist.compression-profile=balanced', '--dist-compression-formats=xz', '--set', 'rust.lld=false', '--disable-dist-src', '--release-channel=nightly', '--enable-debug-assertions', '--enable-overflow-checks', '--enable-llvm-assertions', '--set', 'rust.verify-llvm-ir', '--set', 'rust.codegen-backends=llvm,cranelift,gcc', '--set', 'llvm.static-libstdcpp', '--set', 'gcc.download-ci-gcc=true', '--enable-new-symbol-mangling']
configure: build.build          := x86_64-unknown-linux-gnu
configure: target.x86_64-unknown-linux-gnu.llvm-config := /usr/lib/llvm-19/bin/llvm-config
configure: llvm.link-shared     := True
configure: rust.randomize-layout := True
configure: rust.thin-lto-import-instr-limit := 10
---
---- num::f32::floor stdout ----
---- num::f32::floor stderr ----

thread 'main' panicked at coretests/tests/num/mod.rs:1041:1:
assertion failed: (0.0 as f32).floor().is_sign_negative()
stack backtrace:
   0: __rustc::rust_begin_unwind
   1: core::panicking::panic_fmt
   2: core::panicking::panic
   3: coretests::num::f32::floor
---
---- num::f64_const::floor stdout ----
---- num::f64_const::floor stderr ----

thread 'main' panicked at coretests/tests/num/mod.rs:1077:1:
assertion failed: const { (0.0 as f64).floor().is_sign_negative() }
stack backtrace:
   0: __rustc::rust_begin_unwind
   1: core::panicking::panic_fmt
   2: core::panicking::panic
   3: coretests::num::f64_const::floor
---
---- num::f32_const::floor stdout ----
---- num::f32_const::floor stderr ----

thread 'main' panicked at coretests/tests/num/mod.rs:1053:1:
assertion failed: const { (0.0 as f32).floor().is_sign_negative() }
stack backtrace:
   0: __rustc::rust_begin_unwind
   1: core::panicking::panic_fmt
   2: core::panicking::panic
   3: coretests::num::f32_const::floor
---
---- num::f64::floor stdout ----
---- num::f64::floor stderr ----

thread 'main' panicked at coretests/tests/num/mod.rs:1065:1:
assertion failed: (0.0 as f64).floor().is_sign_negative()
stack backtrace:
   0: __rustc::rust_begin_unwind
   1: core::panicking::panic_fmt
   2: core::panicking::panic
   3: coretests::num::f64::floor
---
    num::f64_const::floor

test result: FAILED. 1902 passed; 4 failed; 4 ignored; 0 measured; 0 filtered out; finished in 1.29s

error: test failed, to rerun pass `-p coretests --test coretests`
env -u RUSTC_WRAPPER CARGO_ENCODED_RUSTDOCFLAGS="-Csymbol-mangling-version=v0\u{1f}-Zrandomize-layout\u{1f}-Zunstable-options\u{1f}--check-cfg=cfg(bootstrap)\u{1f}--check-cfg=cfg(llvm_enzyme)\u{1f}-Dwarnings\u{1f}-Wrustdoc::invalid_codeblock_attributes\u{1f}--crate-version\u{1f}1.89.0-nightly\t(8f5a570ef\t2025-05-25)" CARGO_ENCODED_RUSTFLAGS="-Csymbol-mangling-version=v0\u{1f}-Zrandomize-layout\u{1f}-Zunstable-options\u{1f}--check-cfg=cfg(bootstrap)\u{1f}--check-cfg=cfg(llvm_enzyme)\u{1f}-Zmacro-backtrace\u{1f}-Csplit-debuginfo=off\u{1f}-Clink-arg=-L/usr/lib/llvm-19/lib\u{1f}-Cllvm-args=-import-instr-limit=10\u{1f}-Clink-args=-Wl,-z,origin\u{1f}-Clink-args=-Wl,-rpath,$ORIGIN/../lib\u{1f}-Alinker-messages\u{1f}--cap-lints=allow\u{1f}--cfg\u{1f}randomized_layouts" RUSTC="/checkout/obj/build/x86_64-unknown-linux-gnu/stage1-tools/cg_clif/dist/rustc-clif" RUSTDOC="/checkout/obj/build/x86_64-unknown-linux-gnu/stage1-tools/cg_clif/dist/rustdoc-clif" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/cargo" "test" "--manifest-path" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage1-tools/cg_clif/build/sysroot_tests/Cargo.toml" "--target-dir" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage1-tools/cg_clif/build/sysroot_tests_target" "--locked" "--target" "x86_64-unknown-linux-gnu" "-p" "coretests" "-p" "alloctests" "--tests" "--" "-q" exited with status ExitStatus(unix_wait_status(25856))
Command has failed. Rerun with -v to see more details.
Build completed unsuccessfully in 0:24:26
  local time: Sun May 25 19:49:46 UTC 2025
  network time: Sun, 25 May 2025 19:49:46 GMT
##[error]Process completed with exit code 1.
Post job cleanup.
```

</details>

RalfJung commented:
> First one is: should the rounding methods on f16 and f128 be gated behind #[rustc_const_unstable(feature = {"f16", "f128"}, issue = "116909")] (from https://github.com/rust-lang/rust/issues/116909) or #[rustc_const_unstable(feature = "const_float_round_methods", issue = "141555")]? I think it should be the former one, but I'm not sure.

Good question! The typical thing we do is to gate it behind the `f16`/`f128` feature, but then also have a second *commented out* more specific feature gate:
```rust
#[rustc_const_unstable(feature = "f16", issue = "116909")]
// #[rustc_const_unstable(feature = "const_float_round_methods", issue = "141555")]
pub const fn trunc(self) -> self { ... }
```

> The second question: I added #![feature(const_float_round_methods)] to [library/std/src/lib.rs](https://github.com/rust-lang/rust/pull/141521/commits/805077f140be85968f33050bc1038eb620666e80#diff-317b30ba653f20600042781487d517ae7d07547466b074d09d521f0b9b6ccc52), which was suggested by the compiler itself.

Yes, that's right.
Thanks for asking for confirmation. :)

RalfJung commented:
@rustbot author

RalfJung commented on library/coretests/tests/num/mod.rs:
```diff
@@ -924,6 +924,96 @@ macro_rules! test_float {
             }
+            #[test]
+            #[cfg(not(bootstrap))]
+            fn floor() {
```
FWIW we already have `floor` tests in `library/coretests/tests/floats/f*.rs`. We probably shouldn't duplicate tests, but also those tests can't easily access the "test the same thing in const and regular fn" infrastructure that we have here.

@tgross35 how would you like these tests to be organized? Is it okay to move them from `tests/float/f*.rs` to here?

ruancomelli replied:
> We probably shouldn't duplicate tests, but also those tests can't easily access the "test the same thing in const and regular fn" infrastructure that we have here.

But the "test the same thing in `const` and regular" infrastructure is only used for the float tests, so we could move this infra into `library/coretests/tests/floats`, could we not?

> Also this is the wrong PR to move everything around.

Phew! :smile:

tgross35 replied:
As a thought, what if `coretests/tests/floats/mod.rs` (or `coretests/tests/num.rs`, whatever) defined macros that expand to test cases, rather than entire tests? So each float type defines its own test functions that can include type-specific tests, but some of the copy+paste is gone.

E.g.:

```
// coretests/tests/floats/f32.rs

#[test]
fn test_fract() {
    shared_fract_tests!(f32); // tests used by all sizes
    assert_eq!(f32::math::fract(1.3f32), 0.29999995f32); // type-specific tests
}
```

Separately, it seems like it would be useful to have a `#[test_const]` proc macro that duplicates the test and wraps it in a `const` block with redefined `assert*` macros. Since there are a growing number of cases where we want to be able to verify behavior both at CTFE and at runtime.

RalfJung replied:
> But the "test the same thing in const and regular" infrastructure is only used for the float tests, so we could move this infra into library/coretests/tests/floats, could we not?

Yeah long-term that's the right plan IMO.

> Separately, it seems like it would be useful to have a #[test_const] proc macro that duplicates the test and wraps it in a const block with redefined assert* macros. Since there are a growing number of cases where we want to be able to verify behavior both at CTFE and at runtime.

I think something reasonably close can even be done as a regular macro.

@ruancomelli  For this PR, maybe leave a FIXME comment both at the new and the old tests that this needs deduplicating? It'd also be worth opening an issue to track this and reference it from the FIXME.

RalfJung replied:
Or, if the new tests do cover everything the existing `test_floor` have, remove the old ones, and file an issue for figuring out the organization here.

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-19`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/15287801201/job/43001557893) [(plain)](https://github.com/rust-lang/rust/commit/3af13ac42a5b5305795e4626372ae08a24ea96f5/checks/43001557893/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
#18 exporting to docker image format
#18 sending tarball 21.3s done
#18 DONE 27.9s
##[endgroup]
Setting extra environment values for docker:  --env ENABLE_GCC_CODEGEN=1 --env GCC_EXEC_PREFIX=/usr/lib/gcc/
[CI_JOB_NAME=x86_64-gnu-llvm-19]
[CI_JOB_NAME=x86_64-gnu-llvm-19]
debug: `DISABLE_CI_RUSTC_IF_INCOMPATIBLE` configured.
---
sccache: Listening on address 127.0.0.1:4226
##[group]Configure the build
configure: processing command line
configure: 
configure: build.configure-args := ['--build=x86_64-unknown-linux-gnu', '--llvm-root=/usr/lib/llvm-19', '--enable-llvm-link-shared', '--set', 'rust.randomize-layout=true', '--set', 'rust.thin-lto-import-instr-limit=10', '--set', 'build.print-step-timings', '--enable-verbose-tests', '--set', 'build.metrics', '--enable-verbose-configure', '--enable-sccache', '--disable-manage-submodules', '--enable-locked-deps', '--enable-cargo-native-static', '--set', 'rust.codegen-units-std=1', '--set', 'dist.compression-profile=balanced', '--dist-compression-formats=xz', '--set', 'rust.lld=false', '--disable-dist-src', '--release-channel=nightly', '--enable-debug-assertions', '--enable-overflow-checks', '--enable-llvm-assertions', '--set', 'rust.verify-llvm-ir', '--set', 'rust.codegen-backends=llvm,cranelift,gcc', '--set', 'llvm.static-libstdcpp', '--set', 'gcc.download-ci-gcc=true', '--enable-new-symbol-mangling']
configure: build.build          := x86_64-unknown-linux-gnu
configure: target.x86_64-unknown-linux-gnu.llvm-config := /usr/lib/llvm-19/bin/llvm-config
configure: llvm.link-shared     := True
configure: rust.randomize-layout := True
configure: rust.thin-lto-import-instr-limit := 10
---
     |                ^^^^^^^^
     |
    ::: compiler/rustc_const_eval/src/interpret/place.rs:206:1
     |
206  | pub struct PlaceTy<'tcx, Prov: Provenance = CtfeProvenance> {
     | ----------------------------------------------------------- similarly named struct `PlaceTy` defined here
     |
help: a struct with a similar name exists
     |
1004 -         dest: &MPlaceTy<'tcx, M::Provenance>,
1004 +         dest: &PlaceTy<'tcx, M::Provenance>,
```

</details>

ruancomelli commented on library/coretests/tests/num/mod.rs:
Good idea! Done in f8930f467d5e1ce7c25c9a26a03e26de5b323e9b.

ruancomelli replied:
Note that there's a somewhat unexpected result:
```rust
$fassert!((-0.0 as $fty).fract(), 0.0);
$fassert!((-0.0 as $fty).fract().is_sign_positive());
```
My first intuition told me this should be negative-signed, `-0.0`. However, given that the implementation of `fract(x)` is `1 - trunc(x)`, and `trunc(-0.0)` is `-0.0`, the result becomes clear. But a bit surprising at first.

RalfJung commented on compiler/rustc_const_eval/src/interpret/intrinsics.rs:
```diff
@@ -900,4 +997,19 @@ impl<'tcx, M: Machine<'tcx>> InterpCx<'tcx, M> {
+    {
+        let x: F = self.read_scalar(&args[0])?.to_float()?;
+        let res = x.round_to_integral(mode).value;
+        self.write_scalar(res, dest)?;
```
To fix the test failure, you'll have to do NaN adjustments here. Something like this:
```suggestion
        let res = self.adjust_nan(res, &[x]);
        self.write_scalar(res, dest)?;
```

ruancomelli replied:
Thank you!! Fixed in 85056fd6465d99270f946b8740e2c57a2a462b5e.

RalfJung commented on library/coretests/tests/num/mod.rs:
```diff
@@ -924,6 +924,117 @@ macro_rules! test_float {
+                $fassert!((0.5 as $fty).floor(), 0.0);
+                $fassert!((-0.5 as $fty).floor(), -1.0);
+                $fassert!((1.5 as $fty).floor(), 1.0);
+                $fassert!(($max as $fty).floor(), $max);
+                $fassert!(($min as $fty).floor(), $min);
+                $fassert!(($min_pos as $fty).floor(), 0.0);
+                $fassert!((-$min_pos as $fty).floor(), -1.0);
+                $fassert!(($nan as $fty).floor().is_nan());
+                $fassert!(($inf as $fty).floor(), $inf);
+                $fassert!(($neginf as $fty).floor(), $neginf);
```
The way we refer to these consts changed a bit with https://github.com/rust-lang/rust/pull/141571 so this will need a rebase and similar adjustments.

ruancomelli commented:
@rustbot ready

RalfJung commented:
This PR still has 20 commits -- if you squashed something, you didn't actually push it?

tgross35 commented:
Fwiw usually "squash" actually means "fixup", it's not too useful to keep in the messages that you had to do things like "adjust after rebase" and "fix float tests" at some point :) Not that it's much of a problem of course.

RalfJung commented:
@bors r+

bors commented:
:pushpin: Commit f8e97badb26f1436d062b3bfdd4f50adc41b843a has been approved by `RalfJung`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=RalfJung f8e97badb26f1436d062b3bfdd4f50adc41b843a -->
<!-- homu: {"type":"Approved","sha":"f8e97badb26f1436d062b3bfdd4f50adc41b843a","approver":"RalfJung","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->
